### ISSUE TITLE
[BUG]: armhf binaries from cloudsmith are broken

### ISSUE THREAD

**Issue by shawly** [tvheadend/tvheadend#1665]
### Tvheadend Version

v4.3 (development)

### Build Number

4.3-2292\~g9ac61d767\~raspbianbookworm

### Platform

Raspberry Pi OS (Bookworm)

### Architecture

ARMv6 (arm)

### Category

Other..

### Description

When installing on RPi Zero or on RPi B+ models the tvheadend binary fails with the error "Illegal instructions".
This happens for bookworm and buster.

The packages from the raspberrypi.com mirrors do work though!

**Comment by Flole998** [comment]
Looking at the small package size this likely isn't resolved yet.

**Comment by MichaIng** [comment]
Just for reference: https://tvheadend.org/d/8554-cloudsmith-raspbian-packages-are-armv6-incompatible

**Comment by Flole998** [comment]
Ffmpeg should not be used at that point in the startup. there has to be another issue somewhere. Anyways, the armv6l-packages can now be used for playing around, and if things are resolved they can be removed again

**Comment by MichaIng** [comment]
I just tested the `tvheadend-armv6l` package, and it still throws the illegal instruction error:
```console
root@Zero:~# apt install tvheadend-armv6l
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Recommended packages:
  xmltv-util python python-requests
The following NEW packages will be installed:
  tvheadend-armv6l
0 upgraded, 1 newly installed, 0 to remove and 2 not upgraded.
Need to get 0 B/15.2 MB of archives.
After this operation, 36.1 MB of additional disk space will be used.
debconf: delaying package configuration, since apt-utils is not installed
(Reading database ... 18689 files and directories currently installed.)
Preparing to unpack .../tvheadend-armv6l_4.3-2298~ge287b2fc6~raspbianbookworm_armhf.deb ...
Unpacking tvheadend-armv6l (4.3-2298~ge287b2fc6~raspbianbookworm) ...
Setting up tvheadend-armv6l (4.3-2298~ge287b2fc6~raspbianbookworm) ...
root@Zero:~# tvheadend -v
Illegal instruction
```
I'll try to fix this. Do you know whether it is possible to pass multiple `CFLAGS` via `AUTOBUILD_CONFIGURE_EXTRA` here?
https://github.com/tvheadend/tvheadend/blob/master/Autobuild/armv6l.sh#L1

`armel` is wrong, but is has no effect anyway, as the `ARCH` variable is only effective for cross-compiling and/or Linux builds, right? However, I see here that theoretically, `--cflags=` can be used to pass through `CFLAGS`:
https://github.com/tvheadend/tvheadend/blob/master/configure#L111

But since this `AUTOBUILD_CONFIGURE_EXTRA` is passed through multiple syntax, finally applied to debhelper here, I have no idea how it deals with spaces, quotes and such:
https://github.com/tvheadend/tvheadend/blob/master/debian/rules#L12

Furthermore, I do not understand how these are finally parsed in `configure`: https://github.com/tvheadend/tvheadend/blob/master/configure#L95-L101
```bash
opt=
val=
for opt do
  val=${opt#*=}
  opt=${opt%=*}
  opt=${opt#*--}
  case "$opt" in
```
First it sets `opt` empty, then starts a loop, but without any iteration. So if this is no hidden bash syntax, then this whole loop is a noop. I would usually expect this;
```diff
-for opt do
+for opt in "$@" do
  val=${opt#*=}
- opt=${opt%=*}
+ opt=${opt%%=*}
  opt=${opt#*--}
  case "$opt" in
```
So you let `opt` actually loop through all input arguments. And you allow `=` characters in the value, by using only the part before the first `=` as key, instead of the part before the last `=`. So
```sh
'--cflags=-march=armv6zk -mcpu=arm1176jzf-s -mfloat-abi=hard -mfpu=vfp'
```
results in:
```sh
opt='cflags'
val='-march=armv6zk -mcpu=arm1176jzf-s -mfloat-abi=hard -mfpu=vfp'
```
instead of currently
```sh
opt='cflags=-march=armv6zk -mcpu=arm1176jzf-s -mfloat-abi=hard -mfpu'
val='-march=armv6zk -mcpu=arm1176jzf-s -mfloat-abi=hard -mfpu=vfp'
```
Or do I misunderstand the syntax or how the `--cflags` option in `AUTOBUILD_CONFIGURE_EXTRA` is intended to be used?

I'll tinker with this on an Odroid XU4, which should have the same issue like the Raspbian container with `uname -m` => `armv7l`, resulting in some ARMv7/NEON flags/options.

**Comment by shawly** [comment]
I just had some time to tinker around and I've used docker buildx on my amd64 server to generate an armv6 Docker image for tvheadend which worked quite well, I used the current master branch.
With buildx I can also compile the Debian image variant sucessfully.

The Alpine image does run successfully on my RPi B+, I haven't tested the Debian one yet though.

**Comment by Flole998** [comment]
If you can make alpine packages: Those are currently missing on cloudsmith, so a PR which builds and uploads them would be appreciated.

**Comment by MichaIng** [comment]
> I think you might be right about your change, however on my machine it produces a syntax error. Also I know for sure that the enable-* and disable-* options work, so there is more parsing going on elsewhere, likely in `support/configure.inc`.

Indeed, I missed a `;`, just added it above. However, indeed `for var do` is a short form for `for var in "$@"; do`, that was not aware of, hence it works. So the only change to allow having `=` signs in values, is `opt=${opt%=*}` => `opt=${opt%%=*}`. It removes the longest trailing match instead of the shortest, which makes sense, because for the option, the shortest leading match is removed.

I played around a little, and set
```
AUTOBUILD_CONFIGURE_EXTRA="${AUTOBUILD_CONFIGURE_EXTRA:-} --cflags=-mfpu=vfp"
```
in `armv6l.sh`. Some debug output showed me that indeed `CFLAGS="-mfpu=vfp"` was set with this, but FFmpeg still compiled with
```
gcc ... -mcpu=cortex-a8 -mfpu=neon
```
So the `CFLAGS` variable is not passed to or overwritten in FFmpeg build scripts. I'll have a look.

**Comment by MichaIng** [comment]
So, `Makefile.ffmpeg` unsets/unexports passed C flags. But that is only part of the issue. I narrowed it down to two libraries. First of all, I recognised that the used ARM emulation action does some clever `/proc/cpuinfo` replacement/mount and a QEMU wrapper, so that `uname -m` indeed returns `armv6l`, which it normally does not. But one library checks, if no flags are set explicitly, for each explicitly, a way which returns enabled NEON: https://github.com/pguyot/arm-runner-action/pull/95#issuecomment-1995902056

The `configure` script from [this](https://code.videolan.org/videolan/x264/-/archive/c196240409e4d7c01b47448d93b1f9683aaa7cf7/x264-c196240409e4d7c01b47448d93b1f9683aaa7cf7.tar.bz2) library, used [here](https://github.com/tvheadend/tvheadend/blob/master/Makefile.ffmpeg#L54-L58), contains
```sh
cc_check() {
    if [ -z "$3$4" ]; then
        if [ -z "$1$2" ]; then
            log_check "whether $CC works"
        elif [ -z "$1" ]; then
            log_check "for $2"
        else
            log_check "for $1"
        fi
    elif [ -z "$1" ]; then
        if [ -z "$2" ]; then
            if [ -z "$3" ]; then
                log_check "whether $CC supports $4"
            else
                log_check "whether $CC supports $3"
            fi
        else
            log_check "whether $CC supports $3 with $2"
        fi
    else
        log_check "for $3 in $1";
    fi
    rm -f conftest.c
    for arg in $1; do
        echo "#include <$arg>" >> conftest.c
    done
    if [ -n "$4" ]; then
        echo "$4" >> conftest.c
    fi
    echo "int main (void) { $3 return 0; }" >> conftest.c
    if [ $compiler_style = MS ]; then
        cc_cmd="$CC conftest.c $(cc_cflags $CFLAGS $CFLAGSCLI $CHECK_CFLAGS $2) -link $(cl_ldflags $2 $LDFLAGSCLI $LDFLAGS)"
    else
        cc_cmd="$CC conftest.c $CFLAGS $CFLAGSCLI $CHECK_CFLAGS $2 $LDFLAGSCLI $LDFLAGS -o conftest"
    fi
    if $cc_cmd >conftest.log 2>&1; then
        res=$?
        log_ok
    else
        res=$?
        log_fail
        log_msg "Failed commandline was:"
        log_msg "--------------------------------------------------"
        log_msg "$cc_cmd"
        cat conftest.log >> config.log
        log_msg "--------------------------------------------------"
        log_msg "Failed program was:"
        log_msg "--------------------------------------------------"
        cat conftest.c >> config.log
        log_msg "--------------------------------------------------"
    fi
    return $res
}

...

cc_check '' '' '__asm__("vadd.i16 q0, q0, q0");' && define HAVE_NEON
```
So if seems to write `__asm__("vadd.i16 q0, q0, q0");` to `conftest.c`, tries to compile it, and assumes that NEON is supported, if this succeeds. Probably pguyot has an idea whether/how to have a container with this action, emulating `arm1176 `, failing this test, so that NEON is disabled. But that one can be as well solved by exporting `-mfpu=vfp`, or any `-mcpu` or `-march` flag, since the library then skips the NEON test.

Worse is the 2nd failure caused by https://github.com/webmproject/libvpx: https://github.com/tvheadend/tvheadend/blob/master/Makefile.ffmpeg#L72-L77
- See "all supported target platforms", which does not contain any ARMv6 target: https://github.com/webmproject/libvpx/blob/main/configure#L98
- When it finds the `arm-*-gnueabihf`, it assumes `armv7`: https://github.com/webmproject/libvpx/blob/main/build/make/configure.sh#L777-L782
- And then it enables NEON: https://github.com/webmproject/libvpx/blob/main/build/make/configure.sh#L1006-L1018

This can only be avoided by passing a tuple via `--target` option to this build, but then I am still not sure whether it will work, as the needed target is not in the supported list. In any case, if we want to test fixing this, we must pass the target and/or related build flags from tvheadend builds through `Makefile.ffmpeg` down to at least those two library builds. I can implement just some individual environment variable for this single case, which does not disturb anything else but is just used as flag to set the needed flags/options for these two builds.

**Comment by pguyot** [comment]
@MichaIng my understanding is that `cc_check` checks the compiler can generate neon instructions, not that the target can execute neon instructions. A workaround would be to use a compiler that doesn't generate neon instructions. Within CI, however, you probably use prebuilt gcc binary which is neon enabled.

**Comment by MichaIng** [comment]
I think you are right. Not 100% sure whether a NEON capable CPU is required to compile this code, regardless which compiler is used, but at least it is not additionally executed.

`raspios_lite` from your `arm-runner-action` is used, hence the compiler from Raspbian. While that one is ARMv6-capable (like all of Raspbian), it does not mean necessarily mean that it cannot compile NEON-enabled binaries, at least, obviously, it succeeds compiling this particular code used to check whether NEON is supported or not.

**Comment by Flole998** [comment]
For libvpx, we could just kick it out on armv6. That's fair enough, they don't support it so we remove it. x264 is kinda important though, so I would prefer to keep that one in if possible.

**Comment by MichaIng** [comment]
Agreed, as long as missing `libvpx` does not break too much. h.264 is essential of course.

Out of curiosity, I just tested it on a Raspberry Pi Zero:
```console
root@Zero:~# echo 'int main (void) { __asm__("vadd.i16 q0, q0, q0"); return 0; }' > conftest.c
root@Zero:~# gcc conftest.c -o conftest
/tmp/ccXkkq5a.s: Assembler messages:
/tmp/ccXkkq5a.s:28: Error: selected FPU does not support instruction -- `vadd.i16 q0,q0,q0'
root@Zero:~# gcc conftest.c -mcpu=cortex-a8 -mfpu=neon -o conftest
root@Zero:~# ./conftest
Illegal instruction
```
So indeed it compiles fine with NEON enabled, but fails to execute. So nothing is wrong with the `arm-runner-action` environment in this regards and the Raspbian `gcc` is intended to support NEON.

But NEON indeed needs to be enabled explicitly, at least on an ARMv6 host. This is as well done by the `configure` script of the library, as long as no related flags are passed to the build:
```sh
if [ $asm = auto -a $ARCH = ARM ] ; then
    # set flags so neon is built by default
    [ $compiler == CL ] || echo $CFLAGS | grep -Eq '(-mcpu|-march|-mfpu)' || CFLAGS="$CFLAGS -mcpu=cortex-a8 -mfpu=neon"

    cc_check '' '' '__asm__("add r0, r1, r2");' && define HAVE_ARM_INLINE_ASM
    if [ $compiler = CL ] && cpp_check '' '' 'defined(_M_ARM) && _M_ARM >= 7' ; then
        define HAVE_ARMV6
        define HAVE_ARMV6T2
        define HAVE_NEON
```
So passing `CFLAGS='-march=armv6zk -mcpu=arm1176jzf-s -mfloat-abi=hard -mfpu=vfp'` or a subset would make the test fail. I was originally interpreting if falsely, that the test would be skipped in this case, but it would just fail.

So again we need to pass through `CFLAGS` to at least this build.

**Comment by Flole998** [comment]
How about this: We could extract `-march`, `-mcpu`, `-mfloat-abi` and `-mfpu` if they are specified and apply them again to ffmpeg after doing the undefine. That way specifying those for Tvheadend would make them propagate all the way through to the libraries, right? And armv6 is not the only usecase, when doing builds on x86 you could also enable architecture-specific features if you want/need better performance

**Comment by MichaIng** [comment]
What is actually the reason for unsetting all compiler flags in `Makefile.ffmpeg`? It would be easier and more intuitive if those were just all passed through.

**Comment by Flole998** [comment]
I can't tell you cause that was well before my time, but it was an intentional decision: https://github.com/tvheadend/tvheadend/commit/3bcfe3fa3cccc7df52f2c6f38c3c39719b3cc8f8

You'd have to ask @perexg why that decision was made, maybe he still remembers it. It could very well be something like `-Werror`, which would break ffmpeg and/or it's dependencies.

**Comment by MichaIng** [comment]
I opened a PR to interpret the options/values so that values like `CFLAGS` can contain equal signs: #1687

**Comment by Flole998** [comment]
@MichaIng Has this been resolved by the changes or is this still an issue?

**Comment by MichaIng** [comment]
Reviewing this, indeed the `tvheadend-armv6l` packages is without service and `postinst`. Reason is that the related files do not match the package name. E.g. https://github.com/tvheadend/tvheadend/blob/master/debian/tvheadend.service needs to be renamed to `tvheadend-armv6l.service` to be considered by debhelper. Of course the installed service would then also be named like that.

I think it would be easier to not rename the package, but have only the `armv6l`-compatible one in `https://dl.cloudsmith.io/public/tvheadend/tvheadend/deb/raspbian`, removing everything else. Raspbian is the `armv6l` distro, so if the host is `armv7l` or `aarch64`, no point to use that, but `https://dl.cloudsmith.io/public/tvheadend/tvheadend/deb/debian` instead.

The `-u`/`-g` options however work for me:
```console
root@RPiZeroW:~# tvheadend -u www-data -g www-data -v
tvheadend: version 4.3-2634~geb023c573
```
